metrics-publishers/cloudwatch-metrics-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/CloudWatchMetricsPublisher.java

```diff
@@ -0,0 +1,258 @@
+        this.queue = new LinkedBlockingQueue<>(this.metricQueueSize);
+
+        this.producer = MetricProducer.builder().queue(queue).build();
+        this.producerExecutorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                                                                             .threadNamePrefix("sdk-metrics-cw-producer")
+                                                                             .build());
+
+        this.consumer = MetricConsumer.builder()
+                                      .cloudWatchClient(client)
+                                      .queue(queue)
+                                      .namespace(namespace)
+                                      .build();
+        this.consumerExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder()
+                                                                                      .threadNamePrefix("sdk-metrics-cw-consumer")
+                                                                                      .build());
+    }
+
+    /**
+     * CloudWatch publisher converts the given metrics into {@link MetricDatum} instances and add them
+     * to a queue for publishing.
+     *
+     * @param metricsRegistry registry containing the collected metrics
+     */
+    @Override
+    public void registerMetrics(MetricRegistry metricsRegistry) {
+        try {
+            if (publishStarted.compareAndSet(false, true)) {
+                consumerExecutorService.scheduleAtFixedRate(this::publish,
+                                                            0L,
+                                                            publishFrequency.toMillis(),
+                                                            TimeUnit.MILLISECONDS);
+            }
+
+            producerExecutorService.execute(() -> producer.addMetrics(metricsRegistry));
+        } catch (Throwable throwable) {
+            log.warn(() -> "An error occurred when registering metrics in the publisher", throwable);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> publish() {
+        List<CompletableFuture<PutMetricDataResponse>> futures = new ArrayList<>();
+        int count = 0;
+
+        try {
+            while (queue.peek() != null && ++count <= MAX_SERVICE_CALLS_PER_PUBLISH) {
+                futures.add(consumer.call());
+            }
+        } catch (Throwable throwable) {
+            log.warn(() -> "An error occurred when uploading metrics to CloudWatch.", throwable);
+        }
+
+        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
+    }
+
+    @Override
+    public void close() {
+        try {
+            if (producerExecutorService != null) {
+                producerExecutorService.shutdown();
+            }
+
+            if (client != null) {
```

Comment:
Should we do try-catch for individual resources here so that the failure to close the first resource would not affect the closing of the subsequent resources?

Reply:
Done